src/zql/ivm/graph/operators/full-agg-operators.ts

```diff
@@ -0,0 +1,168 @@
+  #lastOutput: AggregateOut<V, AggregateResult> | undefined;
+
+  constructor(
+    input: DifferenceStreamReader<V>,
+    output: DifferenceStreamWriter<AggregateOut<V, AggregateResult>>,
+    fn: (
+      collection: Multiset<V>,
+      last: AggregateOut<V, AggregateResult> | V,
+    ) => AggregateOut<V, AggregateResult>,
+  ) {
+    const inner = (
+      collection: Multiset<V>,
+    ): Multiset<AggregateOut<V, AggregateResult>> => {
+      let last: V | AggregateOut<V, AggregateResult>;
+      if (this.#lastOutput === undefined) {
+        const iter = collection.entries[Symbol.iterator]();
+        const first = iter.next().value;
+        if (!first) {
+          return new Multiset([]);
+        }
+        last = {
+          ...first[0],
+        };
+      } else {
+        last = this.#lastOutput;
+      }
+      const next = fn(collection, last);
+
+      let ret;
+      if (this.#lastOutput !== undefined) {
+        ret = new Multiset([
+          [this.#lastOutput, -1],
+          [next, 1],
+        ]);
+      } else {
+        ret = new Multiset([[next, 1]]);
+      }
+      this.#lastOutput = next;
+      return ret;
+    };
+    super(input, output, inner);
+  }
+}
+
+export class FullCountOperator<
+  V extends object,
+  Alias extends string,
+> extends FullAggregateOperator<V, [[Alias, number]]> {
+  #count = 0;
+
+  constructor(
+    input: DifferenceStreamReader<V>,
+    output: DifferenceStreamWriter<AggregateOut<V, [[Alias, number]]>>,
+    alias: Alias,
+  ) {
+    super(
+      input,
+      output,
+      (
+        collection: Multiset<V>,
+        last: AggregateOut<V, [[Alias, number]]> | V,
+      ): AggregateOut<V, [[Alias, number]]> => {
+        for (const entry of collection.entries) {
+          this.#count += entry[1];
+        }
+        return {
+          ...(last as AggregateOut<V, [[Alias, number]]>),
+          [alias]: this.#count,
+        };
+      },
+    );
+  }
+}
+
+export class FullAvgOperator<
+  V extends object,
+  Field extends keyof V,
+  Alias extends string,
+> extends FullAggregateOperator<V, [[Alias, number]]> {
+  #numElements = 0;
+  #avg = 0;
+
+  constructor(
+    input: DifferenceStreamReader<V>,
+    output: DifferenceStreamWriter<AggregateOut<V, [[Alias, number]]>>,
+    field: Field,
+    alias: Alias,
+  ) {
+    super(
+      input,
+      output,
+      (
+        collection: Multiset<V>,
+        last: AggregateOut<V, [[Alias, number]]> | V,
+      ): AggregateOut<V, [[Alias, number]]> => {
+        let numElements = 0;
+        let sum = 0;
+        for (const entry of collection.entries) {
+          numElements += entry[1];
+          sum += (entry[0][field] as number) * entry[1];
+        }
+
+        if (this.#numElements + numElements === 0) {
+          this.#avg = 0;
+        } else {
+          this.#avg =
+            (this.#avg * this.#numElements + sum) /
+            (this.#numElements + numElements);
+          this.#numElements += numElements;
+        }
```

Comment:
I don't think you need the first branch. The second branch should compute the same result

Reply:
second branch would divide by zero without the first :(